scRNA-seq-advanced/setup/integration-setup.Rmd

```diff
@@ -0,0 +1,402 @@
+                                     # high variance genes
+                                     subset.row = hvg_list
+                                     )
+
+integrated_sce
+```
+- The `reconstructed` assay are corrected expression values. 
+Make sure to not use them for DE though!
+
+- The `corrected` reduced dimensions are the PCs resulting from the corrected expression values. 
+We can calculate UMAP embeddings from the PCs.
+
+
+```{r}
+# Specify the object's `reducedDim` name
+integrated_sce <- scater::runUMAP(integrated_sce, dimred = "corrected")
+integrated_sce
+```
+
+#### The before/after reveal
+
+```{r}
+# Before
+scater::plotReducedDim(combined_sce,
+                       dimred = "UMAP",
+                       colour_by = "batch",
+                       # Some styling to help us see the points:
+                       point_size = 0.5,
+                       point_alpha = 0.4) +
+  ggplot2::ggtitle("Before integration")
+
+
+# After
+scater::plotReducedDim(integrated_sce,
+                       dimred = "UMAP",
+                       colour_by = "batch",
+                       # Some styling to help us see the points:
+                       point_size = 0.5,
+                       point_alpha = 0.4) +
+  ggplot2::ggtitle("After integration with fastmnn")
+
+```
+
+
+
+
+## Integration, again!
+
+There are more integration tools, like `harmony` for example! 
+`harmony` differs from `fastMNN` (the entire algorithm-aside) in a few key ways:
+
+- `harmony` allows you to specify additional covariates (technology, patient, diagnosis, etc.)
+  - This requires slightly more setup - you'd need to pass in a data frame of covariates, not just a batch vector
+- `harmony` does not calculate corrected expression values and only returns a matrix of corrected PCs
+- a bonus key way is that `harmony` is on CRAN but `batchelor` is on bioconductor.
+
+```{r}
+
+# harmony can either start with PCs or with normalized (uncorrected) expression
+# we already have PCs so let's just ask harmony to use those:
+harmony_result <- harmony::HarmonyMatrix(
+      # PCs
+      data_mat  = reducedDim(combined_sce, "PCA"), 
+      # batch information
+      meta_data = colData(combined_sce)$batch,
+      # don't run PCA since we gave PCs in `data_mat` argument
+      do_pca = FALSE
+    )
+
+# If we wanted to start with expression and have harmony calculate PCs for us:
+#harmony::HarmonyMatrix(
+#  data_mat  = logcounts(combined_sce), 
+#  meta_data = colData(combined_sce)$batch
+#)
+
+# The result is just a bunch of PCs!
+head(harmony_result)
+
+# Let's store this directly into our `combined_sce` object, making sure to name it well!
+reducedDim(combined_sce, "harmony_PCA") <- harmony_result
+
+# Calculate UMAP with the right name for PCs:
+# We do NOT want to overwrite the uncorrected UMAP! Make sure to name this differently with `name`
+combined_sce <- scater::runUMAP(combined_sce, dimred = "harmony_PCA", name = "harmony_UMAP")
```

Comment:
> If we remove Harmony like Josh suggested we should show this as well for fastMNN so they see that you can store the corrected UMAP in the original combined SCE object.

Yeah, I'm adding this and then using `combined_sce` moving forward primarily. Push coming soon!


> I also see both sides to including Harmony. 

Same, and I keep going back and forth! I think it's really important to see differences in methods and don't want to bury that. I wonder if this could be a viable middle ground (CC @jashapiro)- we can have some kind of magic result (code in `setup/`?) where we say "btw we integrated these datasets with 1-2 more tools on the side, and look at the differences!" Then, harmony could live in exercises for them to try out another approach. Another idea potential middle ground is that the harmony chunks are not live but all fully fleshed out, so we can move through it all a bit faster to get to the punchline that different methods are, in fact, different.

Reply:
I like the idea that the harmony chunks could be fully fleshed out and we just run them. Let's plan on doing that, until we reconsider as the instruction notebook gets fleshed out. 

We could even write a function that does the harmony integration and adds the results back to the sce and include that as a chunk, but that is a decision to make later...

Reply:
> Let's plan on doing that, until we reconsider

:chefs kiss: